lib/rrd_updates.ml

```diff
@@ -171,27 +171,36 @@ let of_xml input =
     input
 
 let json_of_t t =
-  let buffer = Buffer.create 4096 in
-
-  let do_data row =
-    Printf.bprintf buffer "{t:%Ld,values:[%s]}" row.time
-      (String.concat "," (List.map Utils.f_to_s (Array.to_list row.row_data)))
+  let open Json in
+  let data_record row =
+    record
+      [
+        ("t", int64 row.time)
+      ; ( "values"
+        , array
+          @@ List.map (fun x -> string (Utils.f_to_s x))
+          @@ Array.to_list row.row_data
+        )
+      ]
   in
-
-  Printf.bprintf buffer "{meta: {start:%Ld,step:%Ld,end:%Ld,rows:%d,columns:%d,"
-    t.start_time t.step t.end_time (Array.length t.data) (Array.length t.legend) ;
-  Printf.bprintf buffer "legend:[%s]},"
-    (String.concat ","
-       (List.map (fun x -> "\"" ^ x ^ "\"") (Array.to_list t.legend))
-    ) ;
-  Printf.bprintf buffer "data:[" ;
-  for i = 0 to Array.length t.data - 2 do
-    do_data t.data.(i) ;
-    Printf.bprintf buffer "%s" ","
-  done ;
-  do_data t.data.(Array.length t.data - 1) ;
-  Printf.bprintf buffer "]}" ;
-  Buffer.contents buffer
+  let meta =
+    record
+      [
+        ( "meta"
+        , record
+            [
+              ("start", int64 t.start_time)
+            ; ("step", int64 t.step)
+            ; ("end", int64 t.end_time)
+            ; ("rows", int (Array.length t.data))
+            ; ("columns", int (Array.length t.legend))
+            ; ("legend", array (List.map string @@ Array.to_list t.legend))
```

Comment:
List.map is not tail recursive and can use a very deep stack, possibly overflowing it, previously this was constructed in a streaming way by appending to a buffer.
It also constructs an extra intermediate list that is thrown away. To use less memory try using 'Seq' here:
convert the Array to a seq, use Seq.map and finally converrt the seq to a list: that will avoid the deep stack usage and the unnecessary intermediate list construction. (although it'll still construct and throw away a lot of temporary closures).